src/modules/sasl.c

```diff
@@ -0,0 +1,154 @@
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; either version 2
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*/
+
+#include "base64.h"
+#include "channel.h"
+#include "child.h"
+#include "commands.h"
+#include "core.h"
+#include "core_api.h"
+#include "hashmap.h"
+#include "logging.h"
+#include "mem.h"
+#include "net.h"
+#include "modules.h"
+#include "string_utils.h"
+#include "user.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+int sasl_start_session (Nick *, User *, Chan *, char **);
+int sasl_finish_session (Nick *, User *, Chan *, char **);
+
+extern 
+
+void child_init()
+{
+    AddHook(HOOK_SASL,&sasl_start_session,"sasl_start_session","sasl");
+    AddHook(HOOK_NICKCREATE,&sasl_finish_session,"sasl_finish_session","sasl");
+}
+
+void child_cleanup()
+{}
+
+int sasl_start_session (__unused Nick *nptr, __unused User *uptr, __unused Chan *cptr, char *parv[])
+{
+    char *sender = parv[0] + 1; // remove leading :
+    char *split[10];
+    int i = 0;
+    User *user;
+
+    char *pch = strtok(parv[1], " ");
+
+    while (pch != NULL) {
+        split[i] = pch;
+        pch = strtok(NULL, " ");
+        i++;
+    }
+
+    char *target = split[0];
+    char *uid = split[1];
+    char *command = split[2]; // H/S/C/D
+    char *tail = split[3];
+    int ret;
+    char *authcid, *password;
+
+    char decoded[40];
+
+    // parse
+    if (strcmp(command, "S") == 0) {
+        // We only support PLAIN for now
+        if (strcmp(tail, "PLAIN")) {
+            SendRaw(":%s SASL %s %s M PLAIN", target, sender, uid);
+            SendRaw(":%s SASL %s %s D F", target, sender, uid);
+            return MOD_CONTINUE;
+        }
+        SendRaw(":%s SASL %s %s C +", target, sender, uid);
+    } else if (strcmp(command, "C") == 0) {
+        ret = b64_decode(split[3], (char *)&decoded, 400);
+        authcid = memchr(decoded, '\0', ret);
+        authcid++;
+        password = memchr(authcid, '\0', ret);
+        password++;
+
+        // check that user exists and password matches
+        user = find_user(authcid);
+        if (!user) {
+            SendRaw(":%s SASL %s %s D F", target, sender, uid);
+            return MOD_CONTINUE;
+        } 
+
+        char *md5pass = md5_hash(password);
+        if (Strcmp(md5pass,user->md5_pass)) {
+            // wrong password
+            SendRaw(":%s SASL %s %s D F", target, sender, uid);
+            return MOD_CONTINUE;
+        }
+
+        sync_user(user);
```

Comment:
I don't think this will work. sync_user() is useful only when the user is already present on channels. IIUC this is not the case here (as we are very early in the connection).

Reply:
Not sure why I had added that, removed.